### PR TITLE
`sage -fixdistributions`

### PR DESCRIPTION
**Pull request by mkoeppe**
This is a new maintenance command for adding/updating `# sage_setup: distribution` directives at the top of source files.

Based on a discussion in https://github.com/sagemath/sage/pull/35884#issuecomment-1646657991

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on https://github.com/sagemath/sage/pull/36533 (CI fix)

### CODE REVIEW

**Comment by kwankyu** [comment]
I understand the use case `sage --fixdistributions --set DISTRIBUTION --from-egg-info`. Do we have use cases with other set of switches? Would you list some of them? For instance,
```
sage --fixdistributions --set sagemath-objects 
``` 
is only dangerous.

**Comment by mkoeppe** [comment]
I have added a bit of documentation. The manual versions of the commands are there simply because whenever an automatic mechanism is provided, there should also be a simpler, manual mechanism.

**Comment by mkoeppe** [comment]
I've made it a bit safer.

File: src/doc/en/developer/packaging_sage_library.rst [hunk 1]
```diff
@@ -146,6 +146,30 @@ The source directory of a distribution package, such as
...
+  source files.  Sage provides a tool ``sage --fixdistributions``
+  that assists with this task. For example::
+
+    $ ./sage --fixdistributions --set sagemath-polyhedra \
+         src/sage/geometry/polyhedron/base*.py
+
+  adds or updates the directives in the specified files; and::
+
+    $ ./sage --fixdistributions --set sagemath-polyhedra \
+         src/sage/geometry/polyhedron
```
Comment (kwankyu) [inline src/doc/en/developer/packaging_sage_library.rst:160]: `--add`?

Comment (mkoeppe) [reply src/doc/en/developer/packaging_sage_library.rst:160]: Thanks, fixed

**Comment by kwankyu** [comment]
This 
```
$ sage -fixdistributions --add=sagemath-categories --from-egg-info
Added 'sage_setup: distribution = sagemath-categories' directive in '/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_categories.py'
Added 'sage_setup: distribution = sagemath-categories' directive in '/Users/kwankyu/GitHub/sage-dev/src/sage/categories/additive_groups.py'
Added 'sage_setup: distribution = sagemath-categories' directive in '/Users/kwankyu/GitHub/sage-dev/src/sage/categories/additive_magmas.py'
```
is too verbose and repetitive. How about logging like
```
Added 'sage_setup: distribution = sagemath-categories' directive in 
'/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_categories.py'
'/Users/kwankyu/GitHub/sage-dev/src/sage/categories/additive_groups.py'
...
```
or something close to it.

**Comment by kwankyu** [comment]
Similarly how about this?
```
$ sage --fixdistributions --set sagemath-objects src/sage/typeset/unicode_characters.py
Changed to 'sage_setup: distribution = sagemath-objects' from 'sage_setup: distribution = sagemath-categories' in
'src/sage/typeset/unicode_characters.py' 
```

**Comment by kwankyu** [comment]
Is this meaningful?
```
$ sage --fixdistributions 
/Users/kwankyu/GitHub/sage-dev/src/sage/ext_data: non-package directory
/Users/kwankyu/GitHub/sage-dev/src/sage/all.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_bliss.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_categories.py: file in distribution 'sagemath-categories'
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_coxeter3.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_environment.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_mcqd.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_meataxe.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_objects.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_repl.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_sirocco.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_tdlib.py: file in distribution ''
/Users/kwankyu/GitHub/sage-dev/src/sage/all_cmdline.py: file in distribution ''
```

**Comment by mkoeppe** [comment]
> Is this meaningful?
> 
> ```
> $ sage --fixdistributions 
> /Users/kwankyu/GitHub/sage-dev/src/sage/ext_data: non-package directory
> /Users/kwankyu/GitHub/sage-dev/src/sage/all.py: file in distribution ''
> /Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_bliss.py: file in distribution ''
> /Users/kwankyu/GitHub/sage-dev/src/sage/all__sagemath_categories.py: file in distribution 'sagemath-categories'
> ```

Yes, without `--set` or `--add`, it's intended to print the distribution information

**Comment by kwankyu** [comment]
Then does this
```
$ sage --fixdistributions src/sage/categories/action.pyx 
src/sage/categories/action.pyx: file in distribution ''
```
mean `src/sage/categories/action.pyx ` is included in no distribution? Is that true, isn't it belong to `sagemath-categories`?

Is this documented?

**Comment by kwankyu** [comment]
I will try again after `make pypi-wheels`...

**Comment by kwankyu** [comment]
Now I understand its function
```
$ sage -fixdistributions --set sagemath-objects --from-egg-info
...
$ sage --fixdistributions src/sage/categories/action.pyx 
src/sage/categories/action.pyx: file in distribution 'sagemath-objects'
```

**Comment by kwankyu** [comment]
An alternative suggestion:
```
src/sage/categories/action.pyx: has no distribution 
src/sage/categories/action.pyx: has distribution 'sagemath-objects'
src/sage/categories/action.pyx: added distribution 'sagemath-objects'
src/sage/categories/action.pyx: changed to distribution 'sagemath-objects' from '...'
```
The advantage is that it is short and succinct, and show the affected file first.

By the way, the concept of "empty distribution" (because you mention "nonempty distribution") is confusing. Is this
```
# sage_setup: distribution = 
```
empty distribution? Do we allow that? I think we should not.

**Comment by kwankyu** [comment]
Now I understand why you allowed 
```
$ sage -fixdistributions --set sagemath-objects some-file-or-directory
```
I guess it is because eventually `manifest.in` could be built from the distributions information of the files, so you want to update distribution info of the files manually. The switch `--from-egg-info` allows the flow of distributions info in the other direction. Nice. Am I right?

**Comment by mkoeppe** [comment]
> An alternative suggestion:
> 
> ```
> src/sage/categories/action.pyx: has no distribution 
> src/sage/categories/action.pyx: has distribution 'sagemath-objects'
> src/sage/categories/action.pyx: added distribution 'sagemath-objects'
> src/sage/categories/action.pyx: changed to distribution 'sagemath-objects' from '...'
> ```
> 
> The advantage is that it is short and succinct, and shows the affected file first.

Thanks for the suggestion. Done in 64e50f3dbf

**Comment by mkoeppe** [comment]
> I guess it is because eventually `manifest.in` could be built from the distributions information of the files, so you want to update distribution info of the files manually.

Yes, that's a possible future step

File: src/sage/misc/package_dir.py [hunk 2]
```diff
@@ -223,8 +223,8 @@ def update_distribution(src_file, distribution, *, verbose=False):
             src_lines.insert(0, f'# {directive}\n')
             any_change = True
             if verbose:
-                print("Added 'sage_setup: " f"distribution = {distribution}' "
-                      f"directive in {src_file!r}")
+                print(f"{src_file}: Added 'sage_setup: "
+                      f"distribution = {distribution}' directive")
     if not any_change:
```
Comment (kwankyu) [inline src/sage/misc/package_dir.py:233]: Both capital `Changed` and `Added` or both small?

**Comment by mkoeppe** [comment]
> the concept of "empty distribution" (because you mention "nonempty distribution") is confusing. Is this
> 
> ```
> # sage_setup: distribution = 
> ```
> 
> empty distribution? 

I agree it is poor terminology. The empty string designates a catch-all distribution, likely sagemath-standard.

**Comment by kwankyu** [comment]
Perhaps we may say "empty distribution (string) designates the standard distribution, which is likely to be sagemath-standard"...

You didn't answer, but I guess you don't allow literally
```
# sage_setup: distribution = 
```
but you meant no `# sage_setup` line at all by "empty distribution".

**Comment by kwankyu** [comment]
How about adding in the help
```
By default, 'sage -fixdistributions' shows the distribution of each file. Distribution '' means the standard distribution (sagemath-standard).
```
or something.

**Comment by kwankyu** [comment]
To prevent needless confusion, we may say "package" instead " packages" in 
```
src/sage/libs/coxeter3: ordinary packages (with __init__.py) cannot be split in several distributions ('', 'sagemath-coxeter3')
```

**Comment by mkoeppe** [comment]
> I guess you don't allow literally
> 
> ```
> # sage_setup: distribution = 
> ```
> 
> but you meant no `# sage_setup` line at all by "empty distribution".

It is allowed and is the same as no `# sage_setup` line.
The script creates such lines when using `--set ''`

**Comment by kwankyu** [comment]
I get
```bash
$ sage --fixdistributions src/sage/monoids
src/sage/monoids/all.py: file in distribution ''
src/sage/monoids/automatic_semigroup.py: file in distribution ''
...
src/sage/monoids/trace_monoid.py: file in distribution ''
$ sage --fixdistributions src/sage/monoids/all.py 
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/kwankyu/GitHub/sage-dev/src/sage/misc/package_dir.py", line 548, in <module>
    handle_file(path)
TypeError: handle_file() missing 1 required positional argument: 'file
```
It should work also with a single file. Right?

**Comment by kwankyu** [comment]
> It is allowed and is the same as no `# sage_setup` line. The script creates such lines when using `--set ''`

OK, then.

**Comment by kwankyu** [comment]
```bash
$ sage --fixdistributions --set sagemath-coxeter3 src/sage/libs/coxeter3
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/kwankyu/GitHub/sage-dev/src/sage/misc/package_dir.py", line 546, in <module>
    handle_file(root, file)
  File "/Users/kwankyu/GitHub/sage-dev/src/sage/misc/package_dir.py", line 521, in handle_file
    package_distributions_per_directives[root].add(distribution)
UnboundLocalError: local variable 'distribution' referenced before assignment
```
??

File: src/bin/sage [hunk 3]
```diff
@@ -482,10 +482,13 @@ usage_advanced() {
...
         echo "                         with actual output."
-        echo "  --fiximports <files|dir>"
+        echo "  --fiximports <files|dirs>"
         echo "                      -- Replace imports from sage.PAC.KAGE.all by specific"
         echo "                         imports when sage.PAC.KAGE is an implicit namespace"
         echo "                         package"
+        echo "  --fixdistributions <files|dirs>"
+        echo "                      -- Check or update 'sage_setup: distribution' directives"
+        echo "                         in source files"
     fi
```
Comment (kwankyu) [inline src/bin/sage:492]:
```
Check or update '# sage_setup: distribution' directives"
```
?

**Comment by kwankyu** [comment]
Here failure of applying ci-fixes is a disaster: B&T was cancelled.

**Comment by mkoeppe** [comment]
I canceled it manually to try out the fix pushed to #36533...

**Comment by kwankyu** [comment]
OK.

**Comment by mkoeppe** [comment]
Turns out the fix arrives too late to fix its own application. I'll just merge #36533 here.

**Comment by kwankyu** [comment]
Currently failure of applying ci-fixes results in failure of the B&T. Shouldn't we add `continue-on-error: true`?

**Comment by mkoeppe** [comment]
As a hotfix that would also arrive too late - the workflows are evaluated before the merging happens

**Comment by kwankyu** [comment]
Not just for a hotfix, this incident may happen again, then perhaps we want to recover from it cleanly.

**Comment by mkoeppe** [comment]
Good point.

**Comment by kwankyu** [comment]
Would you rebase?

**Comment by mkoeppe** [comment]
Rebased

**Comment by kwankyu** [comment]
I get
```
$ ./sage --fixdistributions --set all --from-egg-info
sage --fixdistributions: found egg-info of distribution 'sagemath-bliss'
sage --fixdistributions: found egg-info of distribution 'sagemath-coxeter3'
sage --fixdistributions: found egg-info of distribution 'sagemath-mcqd'
sage --fixdistributions: found egg-info of distribution 'sagemath-meataxe'
sage --fixdistributions: found egg-info of distribution 'sagemath-sirocco'
sage --fixdistributions: found egg-info of distribution 'sagemath-tdlib'
sage --fixdistributions: found egg-info of distribution 'sagemath-environment'
src/sage/misc/package_dir.py: changed 'sage_setup: distribution' from 'sagemath-objects' to 'sagemath-environment'
sage --fixdistributions: found egg-info of distribution 'sagemath-categories'
sage --fixdistributions: found egg-info of distribution 'sagemath-repl'
sage --fixdistributions: found egg-info of distribution 'sagemath-objects'
src/sage/misc/package_dir.py: changed 'sage_setup: distribution' from 'sagemath-environment' to 'sagemath-objects'
sage --fixdistributions: checking consistency
src/sage/graphs/bliss_cpp: missing file all__sagemath_bliss.py
src/sage/doctest/tests: missing file all__sagemath_repl.py
```
Note the two lines beginning with `src/sage/misc/package_dir.py: ...`. They seem to conflict to each other.

**Comment by mkoeppe** [comment]
> Note the two lines beginning with `src/sage/misc/package_dir.py: ...`. They seem to conflict to each other.

Yes, some of the MANIFESTs are not clean yet (even in #35095). In `--set all`, the order matters. Sometimes one of the distributions claims a file but then another, listed later, corrects this claim.

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by kwankyu** [comment]
> > Note the two lines beginning with `src/sage/misc/package_dir.py: ...`. They seem to conflict to each other.
> 
> Yes, some of the MANIFESTs are not clean yet (even in #35095). In `--set all`, the order matters. Sometimes one of the distributions claims a file but then another, listed later, corrects this claim.

Yes. This is a useful feature of `--fixdistributions`, detecting conflicts.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**